usual/tls/tls.c

```diff
@@ -742,4 +742,84 @@ tls_close(struct tls *ctx)
 	return (rv);
 }
 
+static bool
+tls_config_keypair_equal(struct tls_keypair *left_tls_keypair, struct tls_keypair *right_tls_keypair)
+{
+  if (!strings_equal(left_tls_keypair->cert_file, right_tls_keypair->cert_file))
+    return false;
+  if (left_tls_keypair->cert_mem && right_tls_keypair->cert_mem &&
+      !(memcmp(left_tls_keypair->cert_mem, right_tls_keypair->cert_mem, left_tls_keypair->cert_len) == 0))
+    return false;
+  if (!strings_equal(left_tls_keypair->key_file, right_tls_keypair->key_file))
+    return false;
+  if (left_tls_keypair->key_mem && right_tls_keypair->key_mem &&
+      !(memcmp(left_tls_keypair->key_mem, right_tls_keypair->key_mem, left_tls_keypair->key_len) != 0))
+    return false;
+  return true;
+}
+
+static bool
+tls_config_keypairs_equal(struct tls_keypair *tls_keypair_left, struct tls_keypair *tls_keypair_right)
+{
+	bool keypair_unchanged;
+	struct tls_keypair *tls_kp_lt, *next_tls_kp_lt, *tls_kp_rt, *next_tls_kp_rt;
+
+	keypair_unchanged = tls_config_keypair_equal(tls_keypair_left, tls_keypair_right);
+
+	if (keypair_unchanged == false) {
+		return false;
+	}
+
+	for (tls_kp_lt = tls_keypair_left; tls_kp_lt != NULL; tls_kp_lt = next_tls_kp_lt) {
+		next_tls_kp_lt = tls_kp_lt->next;
+		for (tls_kp_rt = tls_keypair_right; tls_kp_rt != NULL; tls_kp_rt = next_tls_kp_rt) {
+			next_tls_kp_rt = tls_kp_rt->next;
+
+			keypair_unchanged = tls_config_keypair_equal(tls_kp_lt, tls_kp_rt);
+		}
+	}
```

Comment:
This logic seems a bit weird, and possibly incorrect. I think it makes more sense to iterate over the left and right keypairs lists in tandem (like `zip` in python). And as soon as one of them doesn't match (or one is NULL and the other is not), return false.

Reply:
Aha! It was my _intent_ to iterate over them in tandem and then somehow I wrote this.

Reply:
Untested, but I expected a loop like this:
```
for (tls_kp_lt = tls_keypair_left, tls_kp_rt = tls_keypair_right; tls_kp_lt != NULL && tls_kp_rt != NULL; tls_kp_lt = tls_kp_lt->next, tls_kp_rt = tls_kp_rt->next) {
    //compare lt and rt
}
```

Reply:
Your loop makes a lot more sense. I want to try to shore up my test coverage here first as my implementation seems obviously wrong.

Reply:
I'm satisfied that my test coverage is improved and your implementation works. Please let me know what you think [e7c75b5](https://github.com/libusual/libusual/pull/57/commits/e7c75b5d6e68e289f7eecf7bab9fd8800e4b9686)